### ISSUE TITLE
[BUG] Modifying of dependencies installed from tgz fails

### ISSUE THREAD

**Issue by Danielku15** [npm/cli#6903]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

This is a re-open of https://github.com/npm/cli/issues/1550 as I ran into a similar problem. 

For testing purposes I locally pack first my NPM packages to `.tgz` files to consume them in a test application living in the same repository. I noticed that a very wrong behavior behavior in my tests with the fact that always an old `.tgz` is pulled from somewhere (a cache?) and not the new flavor is used. 

Even doing a `npm cache clean` does not work.  

### Expected Behavior

`npm install --force` should also ignore packages in the cache if they were installed from a local disk `.tgz` not only from remote registries. Or alternatively there should be an alternative option for fully ignoring all packages which are in the cache.

It seems certainly wrong that with an `npm install --force` and old cached version exists. The Docs read: 

> The `-f` or `--force` argument will force npm to fetch remote resources even if a local copy exists on disk.
> (https://docs.npmjs.com/cli/v10/commands/npm-install?v=true)

Well it is technically not a "remote" resource, but I would expect also the cache to be ignored for local files as there is no other option available I am aware of. 

Doing a manual [`npm cache clean --force`](https://github.com/Danielku15/npm-tgz-install/blob/feature/manual-cache/test.sh#L44) appears to be a (nasty) workaround which makes npm pull the new tgz. 

### Steps To Reproduce

A full reproduction can be seen in [This Repository](https://github.com/Danielku15/npm-tgz-install) and this [workflow run](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671). I use Windows and Linux, and lts/* and latest as node versions. 

The workflow is like follows. I linked the related locations in the log of an example run. 

1. I have a [package](https://github.com/Danielku15/npm-tgz-install/blob/main/package/package.json) and [consumer](https://github.com/Danielku15/npm-tgz-install/blob/main/consumer/package.json). The consumer had an [previous version installed before](https://github.com/Danielku15/npm-tgz-install/blob/main/consumer/package-lock.json#L18). 
2. I build a new .tgz by writing some content into a text file which is packed. (see [here](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:7))
3. I try to install with `--force` but [this fails](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:31) with an integrity error as the tgz changed. 
4. I then remove the lock file and try to [install again](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:36). This [works](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:43). 
5. Printing the [installed file](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:45) shows the correct [expected output](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:7). 
6. Then I [modify](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:49) the package again and [pack it](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:67).
7. I delete the node_modules folder and try to reinstall the package again which now [instantly works](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:73) without integrity error. 
8. [Printing the installed file](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524543671/job/17716298853#step:6:75) though shows the old package installed, not the new one. 

And with a `npm cache clean --force` the new file [is picked up](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524619523/job/17716441873#step:7:96
).

### Environment

- npm: 10.1.0 and 9.8.1
- Node.js: v20.8.1 and v18.18.0
- OS Name: ubuntu-22.04 (github runner image version: 20231001.1.0) and windows-2022 (GitHub runner image version 20231002.1.0)
- System Model Name: GitHub Actions Runner
- npm config:
```ini
; node bin location = /opt/hostedtoolcache/node/20.8.1/x[6](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524619523/job/17716441873#step:5:7)4/bin/node
; node version = v20.[8](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524619523/job/17716441873#step:5:9).1
; npm local prefix = /home/runner/work/npm-tgz-install/npm-tgz-install
; npm version = [10](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524619523/job/17716441873#step:5:11).[1](https://github.com/Danielku15/npm-tgz-install/actions/runs/6524619523/job/17716441873#step:7:1).0
; cwd = /home/runner/work/npm-tgz-install/npm-tgz-install
; HOME = /home/runner
; Run `npm config ls -l` to show all defaults.
```

**Comment by reggi** [comment]
Hey @Danielku15 👋 i've been taking a look at this issue I think I understand the issue you're raising here. 

First, Recreating the setup. This script is largely based on creating your repo from scratch but allows me to add a dimension of time.

TLDR; this creates a tarball for package, then updates the tarball but keeps the version.

```
rm -rf ./package
rm -rf ./consumer

mkdir -p package
pushd package > /dev/null
npm init -y > /dev/null
mkdir -p lib
pushd lib > /dev/null
echo "Hello World 1" > dependency.txt
popd > /dev/null
jq -r '.files += ["lib/dependency.txt"]' package.json > temp.json && mv temp.json package.json
ORIGINAL_SHA=$(npm --no-progress publish --dry-run --json -q | jq -r ".integrity")
npm pack
popd > /dev/null

mkdir -p consumer
pushd consumer > /dev/null
npm init -y > /dev/null
npm install ../package/package-1.0.0.tgz --save
jq -r '.packages["node_modules/package"].integrity' package-lock.json
popd > /dev/null

pushd package > /dev/null
pushd lib > /dev/null
echo "Hello World 2" > dependency.txt
popd > /dev/null
rm -rf package-1.0.0.tgz
UPDATED_SHA=$(npm --no-progress publish --dry-run --json -q | jq -r ".integrity")
npm pack
popd > /dev/null

check_sha() {
    local passed_sha=$1
    if [[ $passed_sha == $ORIGINAL_SHA ]]; then
        echo "-----> SHA is the original SHA"
    elif [[ $passed_sha == $UPDATED_SHA ]]; then
        echo "-----> SHA is the updated SHA"
    else
        echo "-----> SHA is neither the original nor the updated SHA"
    fi
}
```

With this setup we can see how different things don't update the sha.

```
# pushd consumer > /dev/null
# echo "npm install --force"
# npm install --force
# check_sha $(jq -r '.packages["node_modules/package"].integrity' package-lock.json)
# popd > /dev/null

# pushd consumer > /dev/null
# echo "npm ci"
# npm ci
# check_sha $(jq -r '.packages["node_modules/package"].integrity' package-lock.json)
# popd > /dev/null
```

These commands `install` and `ci` fail to take in the updated `package-1.0.0.tgz` because the sha in `package-lock.json` and `node_modules/.package-lock.json` pinned the old version. This is expected.

The following commands do update `package-1.0.0.tgz`.

Simply reinstalling the package:

```
# pushd consumer > /dev/null
# echo "reinstall"
# npm install ../package/package-1.0.0.tgz --save
# check_sha $(jq -r '.packages["node_modules/package"].integrity' package-lock.json)
# popd > /dev/null
```

Or removing `node_modules` as well as `package-lock.json` will get rid of the old sha and get the latest.

```
pushd consumer > /dev/null
echo "npm install"
rm -rf node_modules
rm package-lock.json
npm install
check_sha $(jq -r '.packages["node_modules/package"].integrity' package-lock.json)
popd > /dev/null
```

To me this all makes sense. Installing local packages are a rare edge case, and from a tarball is even rarer, if you chose not to use a tarball it wouldn't rely so much on the sha, tarball's inherently are not supposed to really change their contents, kinda the point of a unique archive with a unchangeable sha, vs a symlinked folder which is much more relaxed.

I'm gonna close this out for now, I don't think there's anything the cli needs to change, and the current behavior makes sense to me.

Please correct me if i'm wrong about my understanding of the issue, and if there's a specific failing scenario you'd wish to see not-fail.

**Comment by Danielku15** [comment]
I modified your script to show my confusion:

```bash
#!/bin/bash
rm -rf ./package
rm -rf ./consumer

# Pack Library
echo "Packing Library"
mkdir -p package
pushd package > /dev/null
npm init -y > /dev/null
mkdir -p lib
pushd lib > /dev/null
initial="Hello World 1"
echo $initial > dependency.txt
popd > /dev/null
jq -r '.files += ["lib/dependency.txt"]' package.json > temp.json && mv temp.json package.json
npm pack
popd > /dev/null

# Consume Library First time
echo "Consuming Library First time"
mkdir -p consumer
pushd consumer > /dev/null
npm init -y > /dev/null
npm install ../package/package-1.0.0.tgz --save
after_consume_1=`cat node_modules/package/lib/dependency.txt`

if [[ $initial == $after_consume_1 ]]; then
    echo "-----> 1 PASS: expected '${initial}' got '${after_consume_1}' "
else 
    echo "-----> 1 FAIL: expected '${initial}' got '${after_consume_1}' "
fi
popd > /dev/null

# modify library
echo "modifying library"
pushd package > /dev/null
pushd lib > /dev/null
modified="hello world 2"
echo $modified > dependency.txt
popd > /dev/null
rm -rf package-1.0.0.tgz
npm pack
popd > /dev/null

# Consume Library a second time (fresh)
echo "Installing again (fresh)"
pushd consumer > /dev/null
rm -rf node_modules
(npm install --force && echo "2nd installing worked") || (echo "Install failed, removing lock" && rm package-lock.json && npm install --force && echo "Install after lock worked")
after_consume_2=`cat node_modules/package/lib/dependency.txt`

if [[ $modified == $after_consume_2 ]]; then
    echo "-----> 2 PASS: expected '${modified}' got '${after_consume_2}' "
else 
    echo "-----> 2 FAIL: expected '${modified}' got '${after_consume_2}' "
fi

popd > /dev/null

```

The first test passes, the second one fails. NPM doesn't install the package from the tgz, it uses an old cached version from somewhere. If it would look at the new file, there should be an integrity check error. 

If I understand it correctly: It re-uses the "old" package because the package-lock.json specifies the checksum and integrity, and it finds a matching variant in the cache. But `--force` should tell it to ignore the "local copy" (in the cache). 

As alternative: If I delete the `package-lock.json` before I install the second time, it will not try to find the package in the cache (it doesn't have a checksum + integrity to do so). Hence it will go again to the file and install the new package cleanly. Then I can even use `npm install` without `--force` to get the new tgz. 

My conclusion: `--force` does not force NPM to look at the file again like it would force it to call to the remote registry. You can use it to force NPM fetching from **remote** resources (e.g. npmjs.org), but not from an **external local file path** resource (e.g. `file:../package/package-1.0.0.tgz`) which appears like a bug to me.
